### PR TITLE
add(slb_backend_server): slb backend server resource & data source

### PR DESCRIPTION
**Pull request by FireWoring**
add new backend server

resource_slb_backend_server_test
屏幕快照 2019-07-26 下午5 07 55

data_source_slb_backend_server_test
屏幕快照 2019-07-26 上午11 28 53

### CODE REVIEW

File: alicloud/resource_alicloud_slb_backend_server.go [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+				addDebug(request.GetActionName(), raw)
+				d.SetPartial("backend_servers")
+			}
+		}
+
+		if d.HasChange("backend_servers") {
+			update = true
+		}
+
+		// d.SetPartial("servers")
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_slb_backend_server.go]: i think 222-226 is not need

File: alicloud/resource_alicloud_slb_backend_server.go [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+				if err != nil {
+					if IsExceptedErrors(err, []string{RspoolVipExist}) {
+						return resource.RetryableError(err)
+					}
+					return resource.NonRetryableError(err)
+				}
+				addDebug(request.GetActionName(), raw)
+				return nil
+			})
+			if err != nil {
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_slb_backend_server.go]: check notfoud?

File: alicloud/service_alicloud_slb.go [hunk 3]
```diff
@@ -103,6 +103,28 @@ func (s *SlbService) DescribeSlbServerGroup(id string) (*slb.DescribeVServerGrou
...
+			err = WrapErrorf(Error(GetNotFoundMessage("SlbBackendServers", id)), NotFoundMsg, AlibabaCloudSdkGoERROR)
+		} else {
+			err = WrapErrorf(err, DefaultErrorMsg, id, request.GetActionName(), AlibabaCloudSdkGoERROR)
+		}
+		return nil, err
+	}
+	addDebug(request.GetActionName(), raw)
+	response, _ := raw.(*slb.DescribeLoadBalancerAttributeResponse)
+	if response.LoadBalancerId == "" {
+		err = WrapErrorf(Error(GetNotFoundMessage("Slb", id)), NotFoundMsg, ProviderERROR)
```
Comment (YY-tomorrow) [inline alicloud/service_alicloud_slb.go]: please change Slb to SlbBackendServers

File: alicloud/service_alicloud_slb.go [hunk 4]
```diff
@@ -331,6 +353,35 @@ RETRY:
...
+	if time.Now().After(deadline) {
+		return WrapErrorf(err, WaitTimeoutMsg, id, GetFunc(1), timeout, Null, id, ProviderERROR)
+	}
+	servers := object.BackendServers.BackendServer
+
+	if len(servers) > 0 {
+		for _, s := range servers {
+			for _, server := range items {
+				info := server.(map[string]interface{})
+				if info["server_id"] == s.ServerId {
```
Comment (YY-tomorrow) [inline alicloud/service_alicloud_slb.go]: info没有赋值，这个条件不会成立

File: website/docs/r/slb_backend_server.html.markdown [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+* `server_id` - (Required) A list backend server ID (ECS instance ID).
+* `weight` - (Optional) Weight of the backend server. Valid value range: [0-100]. 
+* `type` - (Optional) Type of the backend server. Valid value ecs, eni. Default to eni.
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `load_balancer_id` - ID of the load balancer.
+* `backend_servers` - A list of instances that have been added in the SLB.
```
Comment (YY-tomorrow) [inline website/docs/r/slb_backend_server.html.markdown]: 缺少id

File: website/docs/r/slb_backend_server.html.markdown [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+layout: "alicloud"
+page_title: "Alicloud: alicloud_slb_backend_server"
+sidebar_current: "docs-alicloud-resource-slb-backend-server"
+description: |-
+  Provides an Application Load Balancer Backend Server resource.
+---
+
+# alicloud\_slb\_backend\_server
+
+Add a group of backend servers (ECS instance) to the Server Load Balancer or remove them from it.
```
Comment (YY-tomorrow) [inline website/docs/r/slb_backend_server.html.markdown]: 补充一个note：Available in 1.35.0+

File: alicloud/resource_alicloud_slb_backend_server.go [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+	weightAndType := make(map[string][]string)
+	for _, server := range object.BackendServers.BackendServer {
+		key := fmt.Sprintf("%d%s%s", server.Weight, COLON_SEPARATED, server.Type)
+		if v, ok := weightAndType[key]; !ok {
+			weightAndType[key] = []string{server.ServerId}
+		} else {
+			v = append(v, server.ServerId)
+			weightAndType[key] = v
+		}
+	}
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_slb_backend_server.go]: 96-105行不需要

File: alicloud/data_source_alicloud_slb_backend_servers_test.go [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+			"backend_servers.#":             "1",
+			"backend_servers.0.instance_id": CHECKSET,
+			"backend_servers.0.weight":      "100",
+			"backend_servers.0.server_type": CHECKSET,
+		}
+	}
+
+	var fakeDnsRecordsMapFunc = func(rand int) map[string]string {
+		return map[string]string{
+			"backend_servers.#": "0",
```
Comment (YY-tomorrow) [inline alicloud/data_source_alicloud_slb_backend_servers_test.go]: check ids

File: alicloud/resource_alicloud_slb_backend_server.go [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+	raw, err := client.WithSlbClient(func(slbClient *slb.Client) (interface{}, error) {
+		return slbClient.AddBackendServers(request)
+	})
+	if err != nil {
+		return WrapErrorf(err, DefaultErrorMsg, "alicloud_slb_backend_servers", request.GetActionName(), AlibabaCloudSdkGoERROR)
+	}
+	addDebug(request.GetActionName(), raw)
+	response, _ := raw.(*slb.AddBackendServersResponse)
+	d.SetId(response.LoadBalancerId + "_defaultServerGroups")
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_slb_backend_server.go]: loadBalancerId已经被slb instance使用了，_defaultServerGroups用于区分两个资源

File: alicloud/resource_alicloud_slb_backend_server.go [hunk 10]
```diff
@@ -0,0 +1,330 @@
...
+		t := k[1]
+		for _, v := range value {
+			s := map[string]interface{}{
+				"server_id": v,
+				"weight":    w,
+				"type":      t,
+			}
+			servers = append(servers, s)
+		}
+	}*/
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_slb_backend_server.go]: please remove 94-119

File: alicloud/resource_alicloud_slb_backend_server.go [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+				})
+				if err != nil {
+					return WrapErrorf(err, DefaultErrorMsg, strings.Split(d.Id(), "_")[0], request.GetActionName(), AlibabaCloudSdkGoERROR)
+				}
+				addDebug(request.GetActionName(), raw)
+				d.SetPartial("backend_servers")
+			}
+		}
+
+		update = true
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_slb_backend_server.go]: i think update is not need

File: alicloud/resource_alicloud_slb_backend_server.go (hunk 2 above)
Comment (YY-tomorrow) [reply alicloud/resource_alicloud_slb_backend_server.go]: 此处反馈，没有notfound错误类型

File: alicloud/resource_alicloud_slb_backend_server_test.go [hunk 12]
```diff
@@ -0,0 +1,479 @@
...
+		return &SlbService{testAccProvider.Meta().(*connectivity.AliyunClient)}
+	})
+	rac := resourceAttrCheckInit(rc, ra)
+
+	testAccCheck := rac.resourceAttrMapUpdateSet()
+
+	rand := acctest.RandIntRange(1000000, 9999999)
+	name := fmt.Sprintf("tf-testAccSlbBackendServersVpc%d", rand)
+	testAccConfig := resourceTestAccConfigFunc(resourceId, name, resourceBackendServerVpcConfigDependence)
+	testAccConfig2 := resourceTestAccConfigFunc(resourceId, name, resourceBackendServerVpcCountConfigDependence)
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_slb_backend_server_test.go]: Dependence模版可以选用一个比较完整的，不需要两个

File: website/docs/r/slb_backend_server.html.markdown (hunk 6 above)
Comment (YY-tomorrow) [reply website/docs/r/slb_backend_server.html.markdown]: sorry，it is 1.53.0+

File: website/docs/r/slb_backend_server.html.markdown (hunk 5 above)
Comment (YY-tomorrow) [reply website/docs/r/slb_backend_server.html.markdown]: 只需要id即可，load_balancer_id，backend_servers可以去掉

File: alicloud/resource_alicloud_slb_backend_server_test.go [hunk 13]
```diff
@@ -0,0 +1,426 @@
...
+			testAccPreCheck(t)
+		},
+
+		// module name
+		IDRefreshName: resourceId,
+		Providers:     testAccProviders,
+		CheckDestroy:  rac.checkResourceDestroy(),
+		Steps: []resource.TestStep{
+			{
+				Config: testAccConfig(map[string]interface{}{
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_slb_backend_server_test.go]: 缺少count配置

File: alicloud/resource_alicloud_slb_backend_server_test.go [hunk 14]
```diff
@@ -0,0 +1,426 @@
...
+				continue
+			}
+			return err
+		}
+		return fmt.Errorf("SLB backend server %s still exist.", rs.Primary.ID)
+	}
+
+	return nil
+}
+
```
Comment (YY-tomorrow) [inline alicloud/resource_alicloud_slb_backend_server_test.go]: testAccCheckSlbBackendServersDestroy 方法可以删掉

File: alicloud/provider.go [hunk 15]
```diff
@@ -249,6 +250,7 @@ func Provider() terraform.ResourceProvider {
 			"alicloud_slb":                         resourceAliyunSlb(),
 			"alicloud_slb_listener":                resourceAliyunSlbListener(),
 			"alicloud_slb_attachment":              resourceAliyunSlbAttachment(),
+			"alicloud_slb_backend_server":          resourceAliyunSlbBackendServers(),
```
Comment (xiaozhu36) [inline alicloud/provider.go]: keep unified on resource name and resource function.

File: alicloud/resource_alicloud_slb_backend_server.go (hunk 9 above)
Comment (xiaozhu36) [reply alicloud/resource_alicloud_slb_backend_server.go]: Use `loadBalancerId ` is fine.

File: alicloud/data_source_alicloud_slb_backend_servers_test.go [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+data "alicloud_zones" "default" {
+	available_resource_creation = "VSwitch"
+}
+data "alicloud_images" "default" {
+  name_regex = "^ubuntu_16.*_64"
+  most_recent = true
+  owners = "system"
+}
+data "alicloud_instance_types" "default" {
+ 	availability_zone = "${data.alicloud_zones.default.zones.0.id}"
```
Comment (xiaozhu36) [inline alicloud/data_source_alicloud_slb_backend_servers_test.go]: Please removed the 76

File: alicloud/data_source_alicloud_slb_backend_servers_test.go [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+resource "alicloud_vpc" "default" {
+  name = "${var.name}"
+  cidr_block = "172.16.0.0/12"
+}
+
+resource "alicloud_vswitch" "default" {
+  name = "${var.name}"
+  vpc_id = "${alicloud_vpc.default.id}"
+  cidr_block = "172.16.0.0/16"
+  availability_zone = "${data.alicloud_zones.default.zones.0.id}"
```
Comment (xiaozhu36) [inline alicloud/data_source_alicloud_slb_backend_servers_test.go]: The availability zone value should come from instance_types datasource

File: alicloud/resource_alicloud_slb_backend_server.go [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+		}
+	}
+
+	idset := schema.NewSet(schema.HashString, nil)
+	for _, server := range instanceSet.List() {
+		info := server.(map[string]interface{})
+		idset.Add(info["server_id"])
+	}
+	return WrapError(slbService.WaitSlbAttribute(strings.Split(d.Id(), "_")[0], idset, DefaultTimeout))
+}
```
Comment (xiaozhu36) [inline alicloud/resource_alicloud_slb_backend_server.go]: The line 292 - 298 can be removed.

**Pull request merged**